mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java

```diff
@@ -288,59 +293,34 @@ private void updateDocs(CRUDOperationContext ctx,
                         try {
                             ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.PRE_CRUD_UPDATE_DOC, ctx, doc.inputDoc);
                             merge.merge(doc.oldDoc, doc.newDoc);
-                            Translator.populateDocHiddenFields(doc.newDoc, md);
+                            Translator.populateCaseInsensitiveHiddenFields(doc.newDoc, md);
                             updateAttemptList.add(doc);
                         } catch (Exception e) {
                             doc.inputDoc.addError(Error.get("update", MongoCrudConstants.ERR_TRANSLATION_ERROR, e));
                         }
                     } else {
                         doc.inputDoc.addError(Error.get("update",
-                                CrudConstants.ERR_NO_FIELD_UPDATE_ACCESS, paths.toString()));
+                                                        CrudConstants.ERR_NO_FIELD_UPDATE_ACCESS, paths.toString()));
                     }
                 }
                 LOGGER.debug("After checks and merge, updating {} docs", updateAttemptList.size());
                 if (!updateAttemptList.isEmpty()) {
-                    BulkWriteOperation bw = collection.initializeUnorderedBulkOperation();
+                    BulkUpdateCtx updateCtx=new BulkUpdateCtx(collection);
                     for (DocInfo doc : updateAttemptList) {
-                        bw.find(new BasicDBObject("_id", doc.oldDoc.get("_id"))).replaceOne(doc.newDoc);
-                        doc.inputDoc.setCRUDOperationPerformed(CRUDOperation.UPDATE);
+                        BulkUpdateCtx.UpdateDoc updateDoc=new BulkUpdateCtx.UpdateDoc(doc.oldDoc,doc.inputDoc);
+                        updateCtx.addDoc(updateDoc,doc.newDoc);
                     }
-                    try {
-                        if (writeConcern == null) {
-                            LOGGER.debug("Bulk updating docs");
-                            bw.execute();
-                        }
-                        else {
-                            LOGGER.debug("Bulk updating docs with writeConcern={} from execution", writeConcern);
-                            bw.execute(writeConcern);
-                        }
-                    } catch (BulkWriteException bwe) {
-                        LOGGER.error("Bulk write exception", bwe);
-                        handleBulkWriteError(bwe.getWriteErrors(), "update", updateAttemptList);
-                    } catch (RuntimeException e) {
-                    } finally {
-                        for (DocInfo doc : updateAttemptList) {
-                            if (!doc.inputDoc.hasErrors()) {
-                                ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.POST_CRUD_UPDATE_DOC, ctx, doc.inputDoc);
-                            }
+                    updateCtx.execute(writeConcern);
```

Comment:
Identifying values won't change, in this context, _id is the only
identifying value, and it is immutable.

I had the same thought while doing this: with a save, concurrent updates
are indistinguishable from serialized updates, one of them wins, and the
result is always consistent. I chose this implementation to notify the side
that lost the race, to force it to reload the latest copy. Maybe that
doesn't really matter?

On Tue, Aug 9, 2016 at 1:16 PM, Alec Henninger <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/299#discussion_r74126280>
> :
>
> > -                            bw.execute();
> > -                        }
> > -                        else {
> > -                            LOGGER.debug("Bulk updating docs with writeConcern={} from execution", writeConcern);
> > -                            bw.execute(writeConcern);
> > -                        }
> > -                    } catch (BulkWriteException bwe) {
> > -                        LOGGER.error("Bulk write exception", bwe);
> > -                        handleBulkWriteError(bwe.getWriteErrors(), "update", updateAttemptList);
> > -                    } catch (RuntimeException e) {
> > -                    } finally {
> > -                        for (DocInfo doc : updateAttemptList) {
> > -                            if (!doc.inputDoc.hasErrors()) {
> > -                                ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.POST_CRUD_UPDATE_DOC, ctx, doc.inputDoc);
> > -                            }
> > +                    updateCtx.execute(writeConcern);
>
> For save requests, it's interesting that we're detecting concurrent mods
> too. Is it because it's possible identifying values could have changed
> between when we retrieved the document to save over and when we actually do
> the save? Are identity fields allowed to be mutable?
>
> I'm trying to decide if clients issuing a save request would want this
> behavior or not (and if they do, maybe they really should be doing an
> update). I guess I expect a save to potentially blow away other concurrent
> updates since I'm passing in the whole entity.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/299/files/4445c9305f6b8f7a6c756db712b5874118e095f2#r74126280>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADgDDSzX_CgyNAB-hL2TMipf0HEcEf0Yks5qeNIdgaJpZM4JfOdd>
> .
>